packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx

```diff
@@ -1,475 +1,633 @@
-        ).toEqual([defaultProps.reference, [6]]);
+
+        await wait(() => {
+            expect(
+                dispatch.mock.calls.filter(
+                    call => call[0].type === CRUD_GET_MATCHING
+                ).length
+            ).toEqual(3);
+            expect(
+                dispatch.mock.calls.filter(
+                    call => call[0].type === CRUD_GET_MANY
+                ).length
+            ).toEqual(1);
+        });
+
+        rerender(
+            <ReferenceArrayInputController
+                {...defaultProps}
+                input={{ value: [5] }}
+                filter={{ permanentFilter: 'bar' }}
+                sort={{ field: 'foo', order: 'ASC' }}
+                perPage={42}
+            >
+                {children}
+            </ReferenceArrayInputController>
+        );
+
+        await wait(() => {
+            expect(
+                dispatch.mock.calls.filter(
+                    call => call[0].type === CRUD_GET_MATCHING
+                ).length
+            ).toEqual(4);
+            expect(
+                dispatch.mock.calls.filter(
+                    call => call[0].type === CRUD_GET_MANY
+                ).length
+            ).toEqual(1);
+        });
     });
 
-    it('should not call crudGetMany when already fetched input value changes', () => {
-        const crudGetMany = jest.fn();
-        const wrapper = shallow(
+    it('should call crudGetMany when input value changes only with the additional input values', async () => {
+        const children = jest.fn(() => <div />);
+
+        const { dispatch, rerender } = renderWithRedux(
+            <ReferenceArrayInputController
+                {...defaultProps}
+                input={{ value: [5] }}
+            >
+                {children}
+            </ReferenceArrayInputController>
+        );
+
+        await wait();
+        expect(dispatch).toHaveBeenCalledWith({
+            type: CRUD_GET_MANY,
+            meta: {
+                resource: 'tags',
+            },
+            payload: { ids: [5] },
+        });
+        rerender(
             <ReferenceArrayInputController
                 {...defaultProps}
                 input={{ value: [5, 6] }}
-                allowEmpty
-                crudGetMany={crudGetMany}
-            />
+            >
+                {children}
+            </ReferenceArrayInputController>
         );
-        expect(crudGetMany.mock.calls[0]).toEqual([
-            defaultProps.reference,
-            [5, 6],
-        ]);
-        wrapper.setProps({ input: { value: [6] } });
-        expect(crudGetMany.mock.calls.length).toEqual(1);
+
+        await wait();
+
+        expect(dispatch).toHaveBeenCalledWith({
+            type: CRUD_GET_MANY,
+            meta: {
+                resource: 'tags',
+            },
+            payload: { ids: [6] },
+        });
     });
 
-    it('should only call crudGetOne and not crudGetMatching when only the record changes', () => {
```

Comment:
I removed this test as the component should not care about record, only about the value